File: web/test/auth.test.js
// web/test/auth.test.js — round-trip + tamper tests for the wallet auth proof
// (web/src/identity/auth.js), mirroring go/internal/identity/auth_test.go. The
// cross-impl byte check against the pairing vector (msg3) lives in the interop
// test, where the channel binding the signature commits to is available.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { hexToBytes } from '@noble/hashes/utils';
import { deriveWallet } from '../src/identity/wallet.js';
import { signAuth, verifyAuth } from '../src/identity/auth.js';

const wallet = deriveWallet(hexToBytes('00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff'));

test('signAuth round-trips through verifyAuth', () => {
  const challenge = crypto.getRandomValues(new Uint8Array(32));
  const sig = signAuth(wallet, challenge);
  assert.equal(sig.length, 64);
  assert.equal(verifyAuth(wallet.address, challenge, sig), true);
});

test('a tampered challenge does not verify', () => {
  const challenge = crypto.getRandomValues(new Uint8Array(32));
  const sig = signAuth(wallet, challenge);
  const bad = challenge.slice();
  bad[0] ^= 0xff;
  assert.equal(verifyAuth(wallet.address, bad, sig), false);
});

test('a tampered signature does not verify', () => {
  const challenge = crypto.getRandomValues(new Uint8Array(32));
  const sig = signAuth(wallet, challenge);
  const bad = sig.slice();
  bad[0] ^= 0xff;
  assert.equal(verifyAuth(wallet.address, challenge, bad), false);
});

test('the wrong wallet does not verify', () => {
  const challenge = crypto.getRandomValues(new Uint8Array(32));
  const sig = signAuth(wallet, challenge);
  const other = deriveWallet(hexToBytes('ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff'));
  assert.equal(verifyAuth(other.address, challenge, sig), false);
});

test('a bad wallet address or wrong-length sig never throws, returns false', () => {
  const challenge = crypto.getRandomValues(new Uint8Array(32));
  assert.equal(verifyAuth('not-base58-0OIl', challenge, new Uint8Array(64)), false);
  assert.equal(verifyAuth(wallet.address, challenge, new Uint8Array(10)), false);
});

File: go/internal/identity/auth_test.go
// go/internal/identity/auth_test.go
package identity

import (
	"bytes"
	"testing"
)

func mustWallet(t *testing.T) *Wallet {
	t.Helper()
	prf := make([]byte, 32)
	for i := range prf {
		prf[i] = byte(i)
	}
	w, err := DeriveWallet(prf)
	if err != nil {
		t.Fatalf("DeriveWallet: %v", err)
	}
	return w
}

func TestSignAuthVerifies(t *testing.T) {
	w := mustWallet(t)
	challenge := []byte("a fresh channel binding")
	sig := w.SignAuth(challenge)
	if len(sig) != 64 {
		t.Fatalf("SignAuth len = %d, want 64", len(sig))
	}
	if err := VerifyAuth(w.Address, challenge, sig); err != nil {
		t.Fatalf("VerifyAuth(valid) = %v, want nil", err)
	}
}

func TestVerifyAuthRejectsTamperedChallenge(t *testing.T) {
	w := mustWallet(t)
	challenge := []byte("the real challenge")
	sig := w.SignAuth(challenge)
	tampered := bytes.Clone(challenge)
	tampered[0] ^= 0xff
	if err := VerifyAuth(w.Address, tampered, sig); err == nil {
		t.Fatal("VerifyAuth(tampered challenge) = nil, want error")
	}
}

func TestVerifyAuthRejectsTamperedSig(t *testing.T) {
	w := mustWallet(t)
	challenge := []byte("the real challenge")
	sig := w.SignAuth(challenge)
	sig[0] ^= 0xff
	if err := VerifyAuth(w.Address, challenge, sig); err == nil {
		t.Fatal("VerifyAuth(tampered sig) = nil, want error")
	}
}

func TestVerifyAuthRejectsWrongWallet(t *testing.T) {
	w := mustWallet(t)
	challenge := []byte("the real challenge")
	sig := w.SignAuth(challenge)

	prf2 := make([]byte, 32)
	for i := range prf2 {
		prf2[i] = byte(255 - i)
	}
	other, err := DeriveWallet(prf2)
	if err != nil {
		t.Fatalf("DeriveWallet: %v", err)
	}
	if err := VerifyAuth(other.Address, challenge, sig); err == nil {
		t.Fatal("VerifyAuth(wrong wallet) = nil, want error")
	}
}

func TestVerifyAuthRejectsBadWalletEncoding(t *testing.T) {
	w := mustWallet(t)
	sig := w.SignAuth([]byte("x"))
	if err := VerifyAuth("not-base58-0OIl", []byte("x"), sig); err == nil {
		t.Fatal("VerifyAuth(bad wallet encoding) = nil, want error")
	}
}
